Add config file

jt05610 opened the pull request:
# Problem

There is no module-level configuration for `pylabrobot`. This results in .log files being created in the current working directory every time `pylabrobot` is imported.

# Solution

Add `pylabrobot.config` module for project configuration. A `Config` object was created that just configures logging for now. A `pylabrobot.ini` file is found in the current directory. If not found, it is searched for in all parent directories. If it still is not found, it gets created at either the project level that contains the .git directory, or the current directory. This gets called whenever `pylabrobot` is imported.

I also made the testing logs collect in a test_logs directory for each directory that has tests by adding a module-level pytest fixture in `conftest.py`

Currently, `.ini` files are supported, however, I wrote it so that .json, .yaml, .toml, etc can quickly be implemented and supported.

The current configuration that is used for pylabrobot is now in `pylabrobot.ini` for this project.

Also, I created `.editorconfig` with the current project formatting style.

# Example config

```ini
[logging]
level = DEBUG
log_dir = .

```

(Equivalent to the current configuration)

# Testing 

* Wrote `config_test.py` and tested reading and writing .ini-formatted 
* Ran existing test suite to make sure this doesn't break it.

rickwierenga commented on pylabrobot/__init__.py:
```diff
@@ -1,15 +1,61 @@
+  logger.setLevel(level)
+
+  now = datetime.datetime.now().strftime("%Y%m%d")
+  # remove file handler if it exists
+  if len(logger.handlers) > 0:
+    logger.handlers.clear()
```
what is the rationale for this?

jt05610 replied:
This lets `pylabrobot.configure` be called after importing to update the project's logger. It comes into play with the testing fixture I put in for sequestering all test logs into test_logs, if I make it:

```python
@pytest.fixture(autouse=True)
def setup_test_config():
  logger = logging.getLogger("pylabrobot")
  print("initial logging handlers", logger.handlers[0])
  configure(TEST_CONFIG)
  print("testing logging handlers", logger.handlers[0])
  yield

```
I get

```
 initial logging handlers <FileHandler ~/PycharmProjects/pylabrobot/pylabrobot/config/pylabrobot-20240605.log (INFO)>
testing logging handlers <FileHandler ~/PycharmProjects/pylabrobot/pylabrobot/config/test_logs/pylabrobot-20240605.log (INFO)>
```

without this check, it would end up with both loggers being used.

rickwierenga replied:
makes sense, thanks for explaining :)

rickwierenga commented on pylabrobot/config/service/writer.py:
```diff
@@ -0,0 +1,18 @@
+
+  @abstractmethod
+  def write(self, w: T, cfg: Config):
+    """Write a Config object."""
+    raise NotImplementedError
```
is raise needed when method is marked as abstarctmethod?

jt05610 replied:
Probably not, I got into the habit from a book I read which uses both to be explicit https://github.com/cosmicpython/code/blob/master/src/allocation/adapters/repository.py

Happy to take the raise out if you prefer!

rickwierenga replied:
is explicitness the only motivation? In that case I think just "saying it once" is clearer. Happy to discuss

rickwierenga commented on pylabrobot/config/service/file.py:
```diff
@@ -0,0 +1,48 @@
+    if reader is None:
+      raise ValueError(f"Unknown file extension: {file_ext}")
+    with open(r, reader.open_mode) as f:
+      return reader.read(f)
```
should this exist in service.file.py?

jt05610 replied:
I named this poorly -- it is intended for reading from files which could be multiple formats. Right now it is just used with the .ini reader, but can quickly be updated to support config with .json or .yaml.

I separated the file opening, then reading so that i can support from any text stream, i.e. for configuration via http instead of from a file. Could definitely simplify if preferred. I write a lot of Go, so am used to using io.Reader/io.Writer.

rickwierenga replied:
Then my question would be why `FileReader` is needed? It seems that the `IniReader` is able to read files by itself. The only exception is `test_file_reader_writer`, where the opened file is passed.

It would make sense to either:
1. allow ConfigReaders to read files, which would complicate reading from http/other sources
2. always have ConfigReaders read from an io stream, but then FileReader will not be a ConfigReader I think

2 is more powerful, but also more code.

rickwierenga commented on pylabrobot/config/__init__.py:
```diff
@@ -0,0 +1,91 @@
+from pylabrobot.config.service.file import MultiReader, FileWriter
+from pylabrobot.config.service.ini_file import IniReader, IniWriter
+
+DEFAULT_READER = MultiReader(
+  reader_map={
+    "ini": IniReader(),
+  }
+)
```
is `FileReader` missing here?

jt05610 replied:
Changed the name of the MultiReader to be MultiFormatFileReader for clarity. Also changing it to DEFAULT_CONFIG_READER

rickwierenga commented on pylabrobot/config/config.py:
```diff
@@ -0,0 +1,53 @@
+  class Logging:
+    """The logging configuration."""
+    level: int = logging.INFO
+    log_dir: Union[str, Path] = Path(".")
```
would it make things easier to just convert to a Path before storing in Config?

jt05610 replied:
Yes, good call. Taking the union out!

rickwierenga commented on pylabrobot/config/config.py:
```diff
@@ -0,0 +1,53 @@
+  "CRITICAL": logging.CRITICAL,
+}
+
+LOG_TO_STRING = {
+  logging.DEBUG: "DEBUG",
+  logging.INFO: "INFO",
+  logging.WARNING: "WARNING",
+  logging.ERROR: "ERROR",
+  logging.CRITICAL: "CRITICAL",
+}
```
perhaps `LOG_TO_STRING = {v: k for k, v in LOG_FROM_STRING.items() }`?

jt05610 replied:
Yes that's a better idea, changed

rickwierenga replied:
need to push? :)

rickwierenga commented on pylabrobot/config/config_test.py:
```diff
@@ -0,0 +1,62 @@
+    )
+
+
+def test_load_config_creates_default():
+  cfg = load_config("test_config", create_default=True,
```
should we assert this test_config does not exist & unlink/warn if it does? It seems this test can easily fail.

jt05610 replied:
yes, fixed

rickwierenga commented on pylabrobot/config/config_test.py:
```diff
@@ -0,0 +1,62 @@
+from pylabrobot.config.service.writer import ConfigWriter
+
+
+@pytest.fixture
+def tmp_dir(tmp_path):
+  return tmp_path
+
+
+@pytest.fixture
+def fake_config(tmp_dir):
+  return Config(
+    logging=Config.Logging(
+      log_dir=tmp_dir / "logs",
+    )
+  )
```
why are these fixtures?

jt05610 replied:
No good reason I guess :) simplified

rickwierenga commented on pylabrobot/conftest.py:
It seems that this file is not imported anywhere, does it run?

rickwierenga replied:
is this still needed?

rickwierenga replied:
I was a noob: https://docs.pytest.org/en/7.1.x/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files

rickwierenga commented:
Compared to your initial PR, I renamed `Loader` to `Reader` and vice versa. Same for `Writer` and `Saver`. I also split up the files into `config.io` and `config.formats` to make the distinction clearer.

It makes a lot more sense to me to have `read` and `write` be methods of the `io`, eg `FileReader` and `FileWriter` instead of `FileLoader` and `FileSaver`. `JsonReader` is now `JsonLoader`.

What do you think?

jt05610 commented:
> Compared to your initial PR, I renamed `Loader` to `Reader` and vice versa. Same for `Writer` and `Saver`. I also split up the files into `config.io` and `config.formats` to make the distinction clearer.
> 
> It makes a lot more sense to me to have `read` and `write` be methods of the `io`, eg `FileReader` and `FileWriter` instead of `FileLoader` and `FileSaver`. `JsonReader` is now `JsonLoader`.
> 
> What do you think?

I like it! makes sense to me

rickwierenga commented:
anything before merging?

two quick questions:
- mind if I take some of your text from this PR for a quick tutorial on the docs site?
- i recently started plr propaganda on twitter. is it ok with you if I share a link to this PR? do you have an account i should @ / how should I give credit?

jt05610 commented:
> anything before merging?
> 
> two quick questions:
> 
>     * mind if I take some of your text from this PR for a quick tutorial on the docs site?
> 
>     * i recently started plr propaganda on twitter. is it ok with you if I share a link to this PR? do you have an account i should @ / how should I give credit?

* Feel free to use any text!
* I made a twitter with @jt05610 as the handle!

rickwierenga commented:
Thank you for contributing this!
